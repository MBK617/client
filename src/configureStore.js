/**
 * @file Configure redux store
 */

import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import createReducer from './reducers';

/**
 * Configures and returns initial redux store
 *
 * @param {object} initialState The initial state of the application
 * 
 * @return {object} A redux store
 * 
 * @memberof module:Main
 *
 */
function configureStore(initialState = {}) {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(
    createReducer(),
    initialState,
    applyMiddleware(sagaMiddleware)
  );

  store.runSaga = sagaMiddleware.run;
  store.injectedReducers = {};
  store.injectedSagas = {};

  return store;
}

export default configureStore;