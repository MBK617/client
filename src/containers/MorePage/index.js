import MorePage from './morePage';
export default MorePage;