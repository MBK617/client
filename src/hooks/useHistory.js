import { useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { createSelector } from 'reselect';
import { popHistory, pushHistory } from 'containers/App/actions';

const selectApp = createSelector(
  state => state.app,
  substate => substate.toJS()
);

const useHistory = () => {
  const { history } = useSelector(selectApp);
  const dispatch = useDispatch();

  const path = useMemo(() => {
    return history.length > 0 ? history[0] : '';
  }, [history]);

  const goTo = (value) => {
    let newPath = value;
    if(newPath.substr(0, 2) === './') {
      newPath = `${path}/${newPath}`
    }
    if(path !== newPath) {
      dispatch(pushHistory(newPath));
    }
  }

  const goBack = () => {
    if(history.length > 1) {
      dispatch(popHistory());
    }
  }

  return {
    history,
    path,
    goTo,
    goBack
  };
}

export default useHistory;