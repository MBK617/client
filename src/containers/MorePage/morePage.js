import React from 'react';

import Navigation from './components/Navigation';
import LoginPage from './components/LoginPage';
import ProfilePage from './components/ProfilePage';
import SettingsPage from './components/SettingsPage';

import useHistory from 'hooks/useHistory';
import useLoginStatus from 'hooks/useLoginStatus';

const MorePage = () => {
  const isLoggedIn = useLoginStatus();
  const { path } = useHistory();

  const routes = [
    {
      name: "Log In",
      path: 'login',
      component: LoginPage,
      icon: 'account-circle',
      show: !isLoggedIn
    },{
      name: "My Profile",
      path: 'profile',
      component: ProfilePage,
      icon: 'account-circle',
      show: isLoggedIn
    },
    {
      name: "Settings",
      path: 'settings', 
      component: SettingsPage,
      icon: 'cog',
      show: isLoggedIn
    },
    {
      name: "Donate",
      icon: 'heart',
      external: 'https://www.paypal.com/donate/?cmd=_donations&business=bostonmbk617@gmail.com&item_name=My%20Brother%27s%20Keeper%20617A%20Boston%20Nonprofit%C2%A0&currency_code=USD&Z3JncnB0=',
      show: true
    }
  ]

  if(path.split('/').length > 1) {
    const localPath = path.split('/')[1]
    for(let route of routes) {
      if(route.path === localPath && route.show) {
        return React.createElement(route.component);
      }
    }
  }
  return <Navigation routes={routes}/>;
}

export default MorePage;