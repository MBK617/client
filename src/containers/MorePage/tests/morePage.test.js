import React from 'react';
import { Provider } from 'react-redux';
import { render, cleanup } from '@testing-library/react-native';
import configureStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';
import { fromJS } from 'immutable';
import useHistory from 'hooks/useHistory';

import MorePage from '..';

jest.mock('hooks/useHistory');
useHistory.mockImplementation(() => ({
  history: ['home'],
  path: 'home'
}));

const initialState = {
  app: fromJS({
    token: '',
  })
};

describe('<MorePage />', () => {
  const sagaMiddleware = createSagaMiddleware();
  let store, rendered;

  const mockStore = (initialState) => ({ 
    ...configureStore([sagaMiddleware])(initialState), 
    injectedReducers: {}, 
    injectedSagas: {}, 
    runSaga: sagaMiddleware.run 
  });
  
  beforeEach(() => {
    store = mockStore(initialState);
    rendered = render(<Provider store={store}><MorePage /></Provider>);
  })

  afterEach(cleanup);

  it('should render correctly', () => {
    const tree = rendered.toJSON();
    expect(tree).toMatchSnapshot();
  });
});