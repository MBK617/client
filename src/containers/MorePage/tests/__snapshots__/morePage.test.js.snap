// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MorePage /> should render correctly 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 3,
          "borderColor": "rgba(0, 0, 0, 0.1)",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginLeft": 15,
            }
          }
        >
          Log In
        </Text>
      </View>
      <Text />
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 3,
          "borderColor": "rgba(0, 0, 0, 0.1)",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Text />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginLeft": 15,
            }
          }
        >
          Donate
        </Text>
      </View>
      <Text />
    </View>
  </View>
</View>
`;
